### PR TITLE
5291: recalculation of dynamic CSS transforms on hover

### PR DESCRIPTION
**Pull request by alexhartstone**
Adds a conditional transform recalculation method to plot api, and calls it in interactions

### CODE REVIEW

File: src/components/fx/hover.js [hunk 1]
```diff
@@ -339,7 +340,8 @@ function _hover(gd, evt, subplot, noHoverEvent) {
 
             xpx = evt.clientX - dbb.left;
             ypx = evt.clientY - dbb.top;
-
+            
+            plotApi.recalculateTransformInverseIfNecessary(gd);
```
Comment (archmoj) [inline src/components/fx/hover.js:344]: I am not sure we we could attach this function to the `Plotly.js` API.
Instead you could attach it to `gd._fullLayout._recalcTrans...` and call it during interactions.

File: src/plot_api/plot_api.js [hunk 2]
```diff
@@ -3856,9 +3852,25 @@ function makePlotFramework(gd) {
         .style('top', '0px')
         .style('right', '0px');
 
+    fullLayout._lastBBox = gd.getBoundingClientRect();
+
     gd.emit('plotly_framework');
 }
 
+// determines if the graph div requires a recalculation of its inverse matrix transforms by comparing old + new bounding boxes.
+function recalculateTransformInverseIfNecessary(gd, newBBox = null) {
```
Comment (archmoj) [inline src/plot_api/plot_api.js:3861]: Declaring an optional argument e.g. `newBBox` is not supported in ES5.
Please run the command below before commits
```npm run test-syntax```
Or alternatively enable `eslint` in your editor (e.g. vs-code) to catch similar errors.
Thanks.

**Review by archmoj — COMMENTED** [review]
Thanks very much for the follow up PR.
Please find my comments below.

File: src/components/fx/hover.js (hunk 1 above)
Comment (alexhartstone) [reply src/components/fx/hover.js:344]: I did wonder about that. Will do, thanks :)

**Comment by alexhartstone** [comment]
Currently changing tests slightly to check the changes work. Having some issues with the polar tests failing when they shouldn't, will investigate.

File: src/plot_api/plot_api.js [hunk 3]
```diff
@@ -3706,17 +3706,29 @@ function purge(gd) {
...
 function makePlotFramework(gd) {
     var gd3 = d3.select(gd);
     var fullLayout = gd._fullLayout;
-    if(fullLayout._inverseTransform === undefined) {
-        var m = fullLayout._inverseTransform = Lib.inverseTransformMatrix(Lib.getFullTransformMatrix(gd));
-        fullLayout._inverseScaleX = Math.sqrt(m[0][0] * m[0][0] + m[0][1] * m[0][1] + m[0][2] * m[0][2]);
-        fullLayout._inverseScaleY = Math.sqrt(m[1][0] * m[1][0] + m[1][1] * m[1][1] + m[1][2] * m[1][2]);
-    }
+
+    fullLayout._recalculateTransformInverseIfNecessary = recalculateTransformInverseIfNecessary;
```
Comment (archmoj) [inline src/plot_api/plot_api.js:3730]: Let's reduce the key name here:
```js
fullLayout._calcInverseTransform = ...
```
And use it in other places.
Thanks!

File: src/lib/dom.js [hunk 4]
```diff
@@ -150,6 +150,16 @@ function isTransformableElement(element) {
     return element && (element instanceof Element || element instanceof HTMLElement);
 }
 
+function domRectsAreEqual(a, b) {
+    return a && b
```
Comment (archmoj) [inline src/lib/dom.js:154]: Whereas semicolon is optional in JavaScript,
a single line `return a && b` may read as `return a && b;`.
Please consider refactoring as follow:
```js
return (
    a && b &&
    a.x === b.x &&
    a.y === b.y &&
    a.top === b.top &&
    a.left === b.left &&
    a.right === b.right &&
    a.bottom === b.bottom
);
```

Comment (archmoj) [reply src/lib/dom.js:154]: Which also fixes these:
pull-5302-a

**Comment by archmoj** [comment]
Here is a [demo](https://codepen.io/MojtabaSamimi/pen/poEoroV) using the changes made by this branch.
Nice job @alexhartstone :1st_place_medal: 
This looks pretty close if you adjust tests and address syntax errors.
To catch syntax errors before each commit you could simply run the command below:
`npm run test-syntax`

**Comment by alexhartstone** [comment]
Fixing up syntax errors now - npm run test-syntax doesn't seem to be picking up a lot of my syntax errors, weirdly. I've installed eslint and that seems to get it

File: src/lib/svg_text_utils.js [hunk 5]
```diff
@@ -748,6 +748,7 @@ function alignHTMLWith(_base, container, options) {
         var x0 = getLeft() - cRect.left;
         var y0 = getTop() - cRect.top;
         var gd = options.gd || {};
+        gd._fullLayout._calcInverseTransform(gd);
```
Comment (archmoj) [inline src/lib/svg_text_utils.js:751]: `gd` might be undefined considering above line
```js
var gd = options.gd || {};
```
If `gd` could be undefined there, then there is no `gd._fullLayout` and calling `gd._fullLayout._calcInverseTransform(gd);` could cause an error.
So here is my suggestion:
```suggestion
        if(gd._fullLayout && gd._fullLayout._calcInverseTransform) gd._fullLayout._calcInverseTransform(gd);
```

File: src/plot_api/plot_api.js [hunk 6]
```diff
@@ -3706,17 +3706,29 @@ function purge(gd) {
     return gd;
 }
 
+// determines if the graph div requires a recalculation of its inverse matrix transforms by comparing old + new bounding boxes.
+function calcInverseTransform(gd, newBBox) {
```
Comment (archmoj) [inline src/plot_api/plot_api.js:3710]: This function never called using the second argument. So I think you could/should drop it.

**Comment by archmoj** [comment]
Have you tested hover on `mapbox` traces?
You may need to recalculate the transform from `src/plots/mapbox/mapbox.js` considering changes in #5193.

**Comment by archmoj** [comment]
> Have you tested hover on `mapbox` traces?
> You may need to recalculate the transform from `src/plots/mapbox/mapbox.js` considering changes in #5193.

Nevermind. `mapbox` hovers works as expected.

File: src/lib/svg_text_utils.js (hunk 5 above)
Comment (archmoj) [reply src/lib/svg_text_utils.js]: Addressed in https://github.com/plotly/plotly.js/pull/5302/commits/3a5ca3a3281f1b4b0737c09da1ec2cfffa4b0396.

File: src/plot_api/plot_api.js (hunk 6 above)
Comment (archmoj) [reply src/plot_api/plot_api.js]: Fixed in https://github.com/plotly/plotly.js/pull/5302/commits/0b2397ad8627b0f19df3596deebac22de6f798fa.

**Comment by archmoj** [comment]
Thanks @alexhartstone
:dancer:

**Pull request merged**